File: server_asp/src/api/auth/AuthService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using Microsoft.IdentityModel.Tokens;
using shieldtify.common;
using shieldtify.models;
using BC = BCrypt.Net.BCrypt;
namespace shieldtify.api.auth
{
    public class AuthService
    {
        public static APIRes existedEmail(string email)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var existed = dbContext.ClientAccounts.Where(i => i.Email == email).FirstOrDefault() != null;
                return new APIRes(200, existed ? "Email existed" : "Email not existed");
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes sendEmailRegister(string email)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var authToken = new Authenticate
                {
                    Uid = Guid.NewGuid(),
                    Token = Guid.NewGuid().ToString(),
                    UsedTo = "create-account",
                };
                dbContext.Authenticates.Add(authToken);
                dbContext.SaveChanges();
                // generate jwt token
                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes("45b3ecf11da3223addf73ea73912512701ce29d5171451f084ab1ddae1064ea85c5209ba69c190cf8223f58fd771e5a0aaff922be094f7cb101f4aac1c1a1986");
                var tokenDescriptor = new SecurityTokenDescriptor
                {
                    Subject = new ClaimsIdentity(new Claim[]
                    {
                        new Claim("email", email),
                        new Claim("token", authToken.Token),
                    }),
                    Expires = DateTime.UtcNow.AddMinutes(30),
                    SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key),
                        SecurityAlgorithms.HmacSha512Signature),
                };
                var token = tokenHandler.CreateToken(tokenDescriptor);
                var emailService = new EmailService();
                emailService.SendMailForCreatePassword(email, tokenHandler.WriteToken(token)).Wait();
                return new APIRes(200, "Send email successfully");
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes register(string token, string username, string password, string displayname)
        {
            try
            {
                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes("45b3ecf11da3223addf73ea73912512701ce29d5171451f084ab1ddae1064ea85c5209ba69c190cf8223f58fd771e5a0aaff922be094f7cb101f4aac1c1a1986");
                tokenHandler.ValidateToken(token, new TokenValidationParameters
                {
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = new SymmetricSecurityKey(key),
                    ValidateIssuer = false,
                    ValidateAudience = false,
                    ClockSkew = TimeSpan.Zero,
                }, out SecurityToken validatedToken);
                var jwtToken = (JwtSecurityToken)validatedToken;
                var email = jwtToken.Claims.First(x => x.Type == "email").Value;
                var tokenValue = jwtToken.Claims.First(x => x.Type == "token").Value;
                using var dbContext = new ShieldtifyContext();
                var authToken = dbContext.Authenticates.Where(i => i.Token == tokenValue).FirstOrDefault();
                if (authToken == null)
                    return new APIRes(400, "Token is invalid");
                if (authToken.UsedTo != "create-account")
                    return new APIRes(400, "Token is invalid");
                var existed = dbContext.ClientAccounts.Where(i => i.Email == email).FirstOrDefault() != null;
                if (existed)
                    return new APIRes(400, "Email existed");
                // hash password
                var salt = BC.GenerateSalt();
                var hashed = BC.HashPassword(password, salt);
                var client = new ClientAccount
                {
                    Uid = Guid.NewGuid(),
                    Username = username,
                    Password = hashed,
                    DisplayName = displayname,
                    Email = email
                };
                // mark token as used
                authToken.IsUsed = true;
                dbContext.ClientAccounts.Add(client);
                dbContext.SaveChanges();
                return new APIRes(200, "Register successfully");
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static APIRes checkToken(string token, string used_to)
        {
            try
            {
                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes("45b3ecf11da3223addf73ea73912512701ce29d5171451f084ab1ddae1064ea85c5209ba69c190cf8223f58fd771e5a0aaff922be094f7cb101f4aac1c1a1986");
                tokenHandler.ValidateToken(token, new TokenValidationParameters
                {
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = new SymmetricSecurityKey(key),
                    ValidateIssuer = false,
                    ValidateAudience = false,
                    ClockSkew = TimeSpan.Zero,
                }, out SecurityToken validatedToken);
                var jwtToken = (JwtSecurityToken)validatedToken;
                if (jwtToken == null)
                    return new APIRes(400, "Token is invalid");
                if (jwtToken.ValidTo < DateTime.UtcNow)
                    return new APIRes(400, "Token is invalid");
                var tokenValue = jwtToken.Claims.First(x => x.Type == "token").Value;
                using var dbContext = new ShieldtifyContext();
                var authToken = dbContext.Authenticates.Where(i => i.Token == tokenValue).FirstOrDefault();
                if (authToken == null)
                    return new APIRes(400, "Token is invalid");
                if (authToken.IsUsed)
                    return new APIRes(400, "Token is invalid");
                if (authToken.UsedTo != used_to)
                    return new APIRes(400, "Token is invalid");
                return new APIRes(200, "Token is valid");
            }
            catch (System.Exception)
            {
                throw;
            }
        }
        public static APIRes login(string password, string username, string email)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                ClientAccount? user = null;
                if (username != "")
                    user = dbContext.ClientAccounts.Where(i => i.Username == username).FirstOrDefault();
                else if (email != "")
                    user = dbContext.ClientAccounts.Where(i => i.Email == email).FirstOrDefault();
                else
                    return new APIRes(400, "Username or email is required");
                if (user == null)
                    return new APIRes(400, "Username or email is incorrect");
                if (!BC.Verify(password, user.Password))
                    return new APIRes(400, "Username or email is incorrect");
                // generate jwt token
                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes("45b3ecf11da3223addf73ea73912512701ce29d5171451f084ab1ddae1064ea85c5209ba69c190cf8223f58fd771e5a0aaff922be094f7cb101f4aac1c1a1986");
                var tokenDescriptor = new SecurityTokenDescriptor
                {
                    Subject = new ClaimsIdentity(new Claim[]
                    {
                        new Claim("user_id", user.Uid.ToString()),
                        new Claim("role", "client")
                    }),
                    Expires = DateTime.UtcNow.AddMinutes(30),
                    SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key),
                        SecurityAlgorithms.HmacSha512Signature),
                };
                var token = tokenHandler.CreateToken(tokenDescriptor);
                return new APIRes(200, "Login successfully", tokenHandler.WriteToken(token));
            }
            catch (System.Exception)
            {
                throw;
            }
        }
        public static APIRes loginAdmin(string username, string password)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var user = dbContext.Accounts.Where(i => i.Username == username).FirstOrDefault();
                if (user == null)
                    return new APIRes(400, "Username or password is incorrect");
                if (!BC.Verify(password, user.Password))
                    return new APIRes(400, "Username or password is incorrect");
                // generate jwt token
                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes("45b3ecf11da3223addf73ea73912512701ce29d5171451f084ab1ddae1064ea85c5209ba69c190cf8223f58fd771e5a0aaff922be094f7cb101f4aac1c1a1986");
                var tokenDescriptor = new SecurityTokenDescriptor
                {
                    Subject = new ClaimsIdentity(new Claim[]
                    {
                        new Claim("user_id", user.Uid.ToString()),
                        new Claim("role", "admin")
                    }),
                    Expires = DateTime.UtcNow.AddMinutes(30),
                    SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key),
                        SecurityAlgorithms.HmacSha512Signature),
                };
                var token = tokenHandler.CreateToken(tokenDescriptor);
                return new APIRes(200, "Login successfully", tokenHandler.WriteToken(token));
            }
            catch (System.Exception)
            {
                throw;
            }
        }
        public static APIRes sendEmailResetPassword(string email)
        {
            try
            {
                using var dbContext = new ShieldtifyContext();
                var user = dbContext.ClientAccounts.Where(i => i.Email == email).FirstOrDefault();
                if (user == null)
                    return new APIRes(400, "Email not existed");
                var authToken = new Authenticate
                {
                    Uid = Guid.NewGuid(),
                    Token = Guid.NewGuid().ToString(),
                    UsedTo = "reset-password",
                };
                dbContext.Authenticates.Add(authToken);
                dbContext.SaveChanges();
                // generate jwt token
                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes("45b3ecf11da3223addf73ea73912512701ce29d5171451f084ab1ddae1064ea85c5209ba69c190cf8223f58fd771e5a0aaff922be094f7cb101f4aac1c1a1986");
                var tokenDescriptor = new SecurityTokenDescriptor
                {
                    Subject = new ClaimsIdentity(new Claim[]
                    {
                        new Claim("user_id", user.Uid.ToString()),
                        new Claim("token", authToken.Token),
                    }),
                    Expires = DateTime.UtcNow.AddMinutes(30),
                    SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key),
                        SecurityAlgorithms.HmacSha512Signature),
                };
                var token = tokenHandler.CreateToken(tokenDescriptor);
                var emailService = new EmailService();
                emailService.SendMailForResetPassword(email, tokenHandler.WriteToken(token)).Wait();
                return new APIRes(200, "Send email successfully");
            }
            catch (Exception)
            {
                throw;
            }
        }

        public static APIRes resetPassword(string token, string password)
        {
            try
            {
                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes("45b3ecf11da3223addf73ea73912512701ce29d5171451f084ab1ddae1064ea85c5209ba69c190cf8223f58fd771e5a0aaff922be094f7cb101f4aac1c1a1986");
                tokenHandler.ValidateToken(token, new TokenValidationParameters
                {
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = new SymmetricSecurityKey(key),
                    ValidateIssuer = false,
                    ValidateAudience = false,
                    ClockSkew = TimeSpan.Zero,
                }, out SecurityToken validatedToken);
                var jwtToken = (JwtSecurityToken)validatedToken;
                var tokenValue = jwtToken.Claims.First(x => x.Type == "token").Value;
                var userId = jwtToken.Claims.First(x => x.Type == "user_id").Value;
                using var dbContext = new ShieldtifyContext();
                var authToken = dbContext.Authenticates.Where(i => i.Token == tokenValue).FirstOrDefault();
                if (authToken == null)
                    return new APIRes(400, "Token is invalid");
                if (authToken.IsUsed)
                    return new APIRes(400, "Token is invalid");
                if (authToken.UsedTo != "reset-password")
                    return new APIRes(400, "Token is invalid");
                var user = dbContext.ClientAccounts.Where(i => i.Uid.ToString() == userId).FirstOrDefault();
                if (user == null)
                    return new APIRes(400, "Token is invalid");
                // hash password
                var salt = BC.GenerateSalt();
                var hashed = BC.HashPassword(password, salt);
                user.Password = hashed;
                // mark token as used
                authToken.IsUsed = true;
                dbContext.SaveChanges();
                return new APIRes(200, "Reset password successfully");
            }
            catch (System.Exception)
            {
                throw;
            }
        }
    }
}

File: server_nodejs/src/api/auth/auth.service.js
const db = require('../../models');
const { AppError } = require('../../common/errors/AppError');
const {
    sendMailForCreatePassword,
    sendEmailResetPassword,
} = require('../../common/email');
const uuid = require('uuid');
const bcrypt = require('bcrypt');
const JWT = require('jsonwebtoken');
module.exports = {
    existedEmail: async ({ email }) => {
        try {
            const user = await db.client_account.findOne({
                where: {
                    email,
                },
            });
            if (user) throw new AppError(400, 'Email existed');
            return {
                statusCode: 200,
                message: 'Email not existed',
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    sendEmailRegister: async ({ email }) => {
        try {
            const authToken = await db.authenticate.create({
                uid: uuid.v4(),
                usedTo: 'create-account',
                token: uuid.v4(),
            });
            const token = JWT.sign(
                {
                    email,
                    token: authToken.token,
                },
                process.env.JWT_SECRET_KEY
                    ? process.env.JWT_SECRET_KEY
                    : 'default_secret_key',
                {
                    expiresIn: '30m',
                },
            );
            await sendMailForCreatePassword(email, token);
            return {
                statusCode: 200,
                message: 'Send email successfully',
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    register: async (token, { username, password, displayname }) => {
        try {
            var decodeToken = JWT.verify(
                token,
                process.env.JWT_SECRET_KEY
                    ? process.env.JWT_SECRET_KEY
                    : 'default_secret_key',
            );
            if (!decodeToken) {
                throw new AppError(400, 'Token is invalid');
            }
            const authToken = await db.authenticate.findOne({
                where: {
                    token: decodeToken.token,
                },
            });
            const existedUsername = await db.client_account.findOne({
                where: {
                    username,
                },
            });
            if (existedUsername) {
                throw new AppError(400, 'Username existed');
            }
            const email = decodeToken.email;
            if (!authToken) {
                throw new AppError(400, 'Token is invalid');
            }
            if (authToken.usedTo !== 'create-account') {
                throw new AppError(400, 'Token is invalid');
            }
            if (authToken.isUsed) {
                throw new AppError(400, 'Token is invalid');
            }
            const existedUser = await db.client_account.findOne({
                where: {
                    email,
                },
            });
            if (existedUser) {
                throw new AppError(400, 'Email existed');
            }
            const salt = await bcrypt.genSalt(10);
            const hashPassword = await bcrypt.hash(password, salt);
            await db.client_account.create({
                uid: uuid.v4(),
                username,
                password: hashPassword,
                display_name: displayname,
                email,
            });
            authToken.isUsed = true;
            await authToken.save();
            return {
                statusCode: 200,
                message: 'Register successfully',
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    checkToken: async (token, used_to) => {
        try {
            // catch error if token is invalid
            const decodeToken = JWT.verify(
                token,
                process.env.JWT_SECRET_KEY
                    ? process.env.JWT_SECRET_KEY
                    : 'default_secret_key',
            );
            if (!decodeToken) {
                throw new AppError(400, 'Token is invalid');
            }
            const authToken = await db.authenticate.findOne({
                where: {
                    token: decodeToken.token,
                },
            });
            if (!authToken) {
                throw new AppError(400, 'Token is invalid');
            }
            if (authToken.isUsed) {
                throw new AppError(400, 'Token is invalid');
            }
            if (authToken.usedTo !== used_to) {
                throw new AppError(400, 'Token is invalid');
            }
            return {
                statusCode: 200,
                message: 'Token is valid',
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    login: async ({ username = null, password, email = null }) => {
        try {
            let user;
            if (username) {
                user = await db.client_account.findOne({
                    where: {
                        username,
                    },
                });
            } else if (email) {
                user = await db.client_account.findOne({
                    where: {
                        email,
                    },
                });
            } else {
                throw new AppError(400, 'Username or password is required');
            }
            if (!user) {
                throw new AppError(400, 'Username or password is incorrect');
            }
            const validPassword = await bcrypt.compare(password, user.password);
            if (!validPassword) {
                throw new AppError(400, 'Username or password is incorrect');
            }
            const token = JWT.sign(
                {
                    user_id: user.uid,
                    role: 'client',
                },
                process.env.JWT_SECRET_KEY
                    ? process.env.JWT_SECRET_KEY
                    : 'default_secret_key',
                {
                    expiresIn: '24h',
                },
            );
            return {
                statusCode: 200,
                message: 'Login successfully',
                token,
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    loginAdmin: async ({ username, password }) => {
        try {
            const user = await db.account.findOne({
                where: {
                    username,
                },
            });
            if (!user) {
                throw new AppError(400, 'Username or password is incorrect');
            }
            const validPassword = await bcrypt.compare(password, user.password);
            if (!validPassword) {
                throw new AppError(400, 'Username or password is incorrect');
            }
            const token = JWT.sign(
                {
                    user_id: user.uid,
                    role: 'admin',
                },
                process.env.JWT_SECRET_KEY
                    ? process.env.JWT_SECRET_KEY
                    : 'default_secret_key',
                {
                    expiresIn: '24h',
                },
            );
            return {
                statusCode: 200,
                message: 'Login successfully',
                token,
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    sendEmailResetPassword: async (email) => {
        try {
            const user = await db.client_account.findOne({
                where: {
                    email,
                },
            });
            if (!user) {
                throw new AppError(400, 'Email not existed');
            }
            const authToken = await db.authenticate.create({
                uid: uuid.v4(),
                usedTo: 'reset-password',
                token: uuid.v4(),
            });
            const token = JWT.sign(
                {
                    user_id: user.uid,
                    token: authToken.token,
                },
                process.env.JWT_SECRET_KEY
                    ? process.env.JWT_SECRET_KEY
                    : 'default_secret_key',
                {
                    expiresIn: '30m',
                },
            );
            await sendEmailResetPassword(email, token);
            return {
                statusCode: 200,
                message: 'Send email successfully',
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
    resetPassword: async (token, { password }) => {
        try {
            const decodeToken = JWT.verify(
                token,
                process.env.JWT_SECRET_KEY
                    ? process.env.JWT_SECRET_KEY
                    : 'default_secret_key',
            );
            if (!decodeToken) {
                throw new AppError(400, 'Token is invalid');
            }
            const authToken = await db.authenticate.findOne({
                where: {
                    token: decodeToken.token,
                },
            });
            if (!authToken) {
                throw new AppError(400, 'Token is invalid');
            }
            if (authToken.usedTo !== 'reset-password') {
                throw new AppError(400, 'Token is invalid');
            }
            if (authToken.isUsed) {
                throw new AppError(400, 'Token is invalid');
            }
            const user = await db.client_account.findOne({
                where: {
                    uid: decodeToken.user_id,
                },
            });
            if (!user) {
                throw new AppError(400, 'Token is invalid');
            }
            const salt = await bcrypt.genSalt(10);
            const hashPassword = await bcrypt.hash(password, salt);
            user.password = hashPassword;
            user.changed_password_at = new Date();
            await user.save();
            authToken.isUsed = true;
            await authToken.save();
            return {
                statusCode: 200,
                message: 'Reset password successfully',
            };
        } catch (error) {
            throw new AppError(error.statusCode, error.message);
        }
    },
};
